allennlp/modules/seq2seq_encoders/multi_head_self_attention.py

```diff
@@ -101,33 +93,33 @@ def forward(self,  # pylint: disable=arguments-differ
         """
         num_heads = self._num_heads
 
-        batch_size, timesteps, hidden_dim = inputs.size()
+        batch_size, timesteps, _ = inputs.size()
         if mask is None:
             mask = Variable(inputs.data.new(batch_size, timesteps).fill_(1.0))
 
-        # Treat the queries, keys and values each as a ``num_heads`` size batch.
-        # shape (num_heads, batch_size * timesteps, hidden_dim)
-        inputs_per_head = inputs.repeat(num_heads, 1, 1).view(num_heads,
-                                                              batch_size * timesteps,
-                                                              hidden_dim)
-        # Do the projections for all the heads at once.
-        # Then reshape the result as though it had a
-        # (num_heads * batch_size) sized batch.
-        queries_per_head = torch.bmm(inputs_per_head, self._query_projections)
-        # shape (num_heads * batch_size, timesteps, attention_dim)
-        queries_per_head = queries_per_head.view(num_heads * batch_size,
-                                                 timesteps,
-                                                 self._attention_dim)
-
-        keys_per_head = torch.bmm(inputs_per_head, self._key_projections)
-        # shape (num_heads * batch_size, timesteps, attention_dim)
-        keys_per_head = keys_per_head.view(num_heads * batch_size,
-                                           timesteps,
-                                           self._attention_dim)
-
-        values_per_head = torch.bmm(inputs_per_head, self._value_projections)
-        # shape (num_heads * batch_size, timesteps, attention_dim)
-        values_per_head = values_per_head.view(num_heads * batch_size, timesteps, self._values_dim)
+        # Shape (batch_size, timesteps, 2 * attention_dim + values_dim)
+        combined_projection = self._combined_projection(inputs)
+
+        # Shape (batch_size, timesteps, attention_dim)
+        queries = combined_projection[:, :, :self._attention_dim].contiguous()
+        keys = combined_projection[:, :, self._attention_dim :2 * self._attention_dim].contiguous()
+        # Shape (batch_size, timesteps, values_dim)
+        values = combined_projection[:, :, 2 * self._attention_dim:].contiguous()
+
+        # Shape (num_heads * batch_size, timesteps, values_dim / num_heads)
```

Comment:
I feel like `sequence_length` is a better name here than `timesteps`, but it's already written this way, and it's not a huge deal if you don't want to change it.

And no, I can't think of a better way to do this than what you're doing.  The way you're doing the projection of the inputs basically forces you to do a transpose here.  Maybe you could do the transpose once, before you split the queries, keys and values?  It would change the semantics of the weights in your combined projection (i.e., currently weights for queries, keys, and values are grouped together, but if you do this the weights for the heads will be grouped together), but that should be fine, as they are all just learned parameters that are initialized identically, anyway.

Reply:
Hmm i'm not sure that will save me much, because then i'll have to transpose again anyway to get the embedding dimension as the last dim. I'll leave it as it is for now and optimise it if it's too slow.

Reply:
Oh, I meant do a `.view()` to get four dimensions, and do a `.transpose(1, 2)`, like you're currently doing.  You can just do it once, instead of three times.  It's not a big deal, though, and yeah, optimize later if it's necessary.